Add 'skaffold apply' command (#5543)
GoogleContainerTools/skaffold: docs/content/en/docs/workflows/ci-cd.md

weight: 40
---

Skaffold provides several features and sub-command "building blocks" that make it very useful for integrating with (or creating entirely new) CI/CD pipelines.

> The ability to use the same skaffold.yaml for iterative development and continuous delivery make handing off an application from a development team to an ops team a breeze!

Reviewer: Until we have more feedback and usage, we should hold off on the superlatives.
```suggestion
The ability to use the same `skaffold.yaml` for iterative development and continuous delivery eases handing off an application from a development team to an ops team.
```

Let's start with the simplest use case: a single, full deployment of your application.

## `skaffold run`
{{< maturity "run" >}}

> `skaffold run` is a single command for a one-off deployment. It runs through every major phase of the Skaffold lifecycle: building your application images, tagging these images (and optionally pushing them to a remote registry), deploying your application to the target cluster, and monitoring the created resources for readiness.

Reviewer: Would be worth pulling the lifecycle phase descriptions to the top so that they can be referenced by name here.

Author: i'm keeping a running list of improvements we should make to these docs, but they'll get addressed outside of this PR

We recommend `skaffold run` for the simplest Continuous Delivery setup, where it is sufficient to have a single step that deploys from version control to a cluster.

For more sophisticated Continuous Delivery pipelines, Skaffold offers building blocks:

- [healthcheck]({{<relref "/docs/workflows/ci-cd#waiting-for-skaffold-deployments-using-healthcheck">}}) - 
wait for `deployments` to stabilize and succeed only if all deployments are successful
- [`skaffold build`]({{<relref "/docs/workflows/ci-cd#skaffold-build-skaffold-deploy">}}) - build, tag and push artifacts to a registry
- [`skaffold deploy`]({{<relref "/docs/workflows/ci-cd#skaffold-build-skaffold-deploy">}})  - deploy built artifacts to a cluster
- [`skaffold render`]({{<relref "/docs/workflows/ci-cd#skaffold-render">}})  - export the transformed Kubernetes manifests for GitOps workflows

> <!-- - [`skaffold apply`]() - send hydrated Kubernetes manifests to the API server to create resources on the target cluster -->

Reviewer: Should this be commented?  `apply` is mentioned several places below.

Author: oof, sorry. I was having trouble with hugo not being able to find this reference, let me fix

```

## GitOps-style continuous delivery: `skaffold render` | `skaffold apply`
{{< maturity "apply" >}}

GitOps-based CD pipelines traditionally see fully-hydrated Kubernetes manifests committed to a configuration Git repository (separate from the application source), which triggers a deployment pipeline that applies the changes to resources on the cluster. Skaffold has two built-in commands that enable easy GitOps pipeline workflows - `skaffold render` and `skaffold apply`.

`skaffold render` builds all application images from your artifacts, templates the newly-generated image tags into your Kubernetes manifests (based on your project's deployment configuration), and then prints out the final hydrated manifests to a file or your terminal. This allows you to capture the full, declarative state of your application in configuration rather than actually applying changes to your cluster, and use this configuration in a GitOps pipeline by committing it to a separate Git repository.

> `skaffold apply` will consume one or more fully-hydrated Kubernetes manifests, and send them directly to the Kubernetes control plane via `kubectl` to create resources on the target cluster. After creating the resources on your cluster, `skaffold apply` will use Skaffold's built-in health checking to monitor the created resources for readiness. See [resource health checks]({{<relref "/docs/workflows/ci-cd#waiting-for-skaffold-deployments-using-healthcheck">}}) for more information on how Skaffold's resource health checking works.

Reviewer: ```suggestion
`skaffold apply` consumes one or more fully-hydrated Kubernetes manifests, and then sends the results directly to the Kubernetes control plane via `kubectl` to create resources on the target cluster. After creating the resources on your cluster, `skaffold apply` uses Skaffold's built-in health checking to monitor the created resources for readiness. See [resource health checks]({{<relref "/docs/workflows/ci-cd#waiting-for-skaffold-deployments-using-healthcheck">}}) for more information on how Skaffold's resource health checking works.
```

> *Note: `skaffold apply` will always use `kubectl` to deploy resources to a target cluster, regardless of deployment configuration in the provided skaffold.yaml. A small subset of deploy configuration will be honored when running `skaffold apply`:*

Reviewer: ```suggestion
*Note: `skaffold apply` always uses `kubectl` to deploy resources to a target cluster, regardless of deployment configuration in the provided skaffold.yaml. Only a small subset of deploy configuration is honored when running `skaffold apply`:*
```

* deploy.statusCheckDeadlineSeconds
* deploy.kubeContext
* deploy.logs.prefix
* deploy.kubectl.flags
* deploy.kubectl.defaultNamespace
* deploy.kustomize.flags
* deploy.kustomize.defaultNamespace

{{<alert title="Note">}}

> While `skaffold apply` will attempt to honor the deployment configuration mentioned above, if conflicting configuration is detected in a multi-configuration project, `skaffold apply` will not work.

Reviewer: ```suggestion
`skaffold apply` attempts to honor the deployment configuration mentioned above.  But when conflicting configuration is detected in a multi-configuration project, `skaffold apply` will not work.
```

{{</alert>}}

`skaffold apply` works with any arbitrary Kubernetes YAML, whether it was generated by Skaffold or not, making it an ideal counterpart to `skaffold render`.

### Example: Hydrating Kubernetes resources using `skaffold render`, then sending them to the cluster using `skaffold apply`:

```code

> -> skaffold render --output render.yaml

Reviewer: I'd suggest sticking with the old-skool `$ ` prompt.  I wasn't sure what the `->` indicated.

Author: ah thanks, copied from my terminal and didn't think about it

```
```yaml
# render.yaml
apiVersion: v1
kind: Pod
metadata:
  name: getting-started
  namespace: default
spec:
  containers:

>   - image: skaffold-example:8fd51906a00a028584c9db41765c51672a45263cd564bbd269690a6181fdcd01

Reviewer: This image name doesn't look … right.  At least, I don't think rendering output for a local cluster is really a typical usecase.

Author: updated

  name: getting-started
  namespace: default
spec:
  containers:
  - image: gcr.io/k8s-skaffold/skaffold-example:v0.41.0-57-gbee90013@sha256:eeffb639f53368c4039b02a4d337bde44e3acc728b309a84353d4857ee95c369
  - image: skaffold-example:8fd51906a00a028584c9db41765c51672a45263cd564bbd269690a6181fdcd01
    name: getting-started
```

> We can then pipe this yaml to kubectl:

Reviewer: I think some text is still helpful here.
```
We can later apply this output:
```

Author: added
